src/log-store/src/kafka/log_store.rs

```diff
@@ -48,33 +115,103 @@ impl LogStore for KafkaLogStore {
-        todo!()
+        if entries.is_empty() {
+            return Ok(AppendBatchResponse::default());
+        }
+
+        // The entries are grouped by topic since the number of regions might be very large
+        // while the number of topics are well controlled,
+        let mut topic_entries: HashMap<_, Vec<_>> = HashMap::new();
+        // A utility map used to construct the result response.
+        let mut topic_regions: HashMap<_, Vec<_>> = HashMap::new();
+        for entry in entries {
+            let topic = entry.ns.topic.clone();
+            let region_id = entry.ns.region_id;
+            topic_entries.entry(topic.clone()).or_default().push(entry);
+            topic_regions.entry(topic).or_default().push(region_id);
+        }
+
+        // Appends each group of entries to the corresponding topic.
+        let tasks = topic_entries
+            .into_iter()
+            .map(|(topic, entries)| self.append_batch_to_topic(entries, topic))
+            .collect::<Vec<_>>();
+        let topic_offset: HashMap<_, _> = futures::future::try_join_all(tasks)
+            .await?
+            .into_iter()
+            .collect();
+
+        let mut region_offset = HashMap::new();
+        for (topic, regions) in topic_regions {
+            let offset = topic_offset[&topic];
+            regions.into_iter().for_each(|region| {
+                region_offset.insert(region, offset);
+            });
+        }
+        Ok(AppendBatchResponse {
+            offsets: region_offset,
+        })
     }
 
-    /// Create a new `EntryStream` to asynchronously generates `Entry` with ids
-    /// starting from `id`. The generated entries will be filtered by the namespace.
+    /// Creates a new `EntryStream` to asynchronously generates `Entry` with entry ids
+    /// starting from `entry_id`. The generated entries will be filtered by the namespace.
     async fn read(
         &self,
         ns: &Self::Namespace,
         entry_id: EntryId,
     ) -> Result<SendableEntryStream<Self::Entry, Self::Error>> {
-        todo!()
+        let topic = ns.topic.clone();
+        let region_id = ns.region_id;
+        let offset = try_get_offset(entry_id)?;
+
+        let raw_client = self
+            .client_manager
+            .get_or_insert(&topic)
+            .await?
+            .raw_client
+            .clone();
+        // Reads the entries starting from exactly the specified offset.
+        let mut stream_consumer = StreamConsumerBuilder::new(raw_client, StartOffset::At(offset))
+            .with_max_batch_size(self.config.max_batch_size.as_bytes() as i32)
+            .with_max_wait_ms(self.config.max_wait_time.as_millis() as i32)
+            .build();
+        let stream = async_stream::stream!({
+            while let Some(consume_result) = stream_consumer.next().await {
+                yield handle_consume_result(consume_result, &topic, region_id, offset);
+            }
+        });
```

Comment:
How do we ensure that `EntryOffest` and `EntryId` have the same order?

In other words, the following mapping seems possible.
```
EntryId, EntryOffest
1, 10
2, 8
3, 5,
4, 1
5, 20
```

Then, the log will lose the correct order.

Reply:
We will ensure that the order the entries being written into the WAL and the order they being written into the memtable are identical.